plugins/http_plugin/include/eosio/http_plugin/beast_http_listener.hpp

```diff
@@ -121,7 +121,9 @@ class beast_http_listener : public std::enable_shared_from_this<beast_http_liste
                fail(ec, "accept", self->plugin_state_->logger, "closing connection");
             } else {
                // Create the session object and run it
-               std::string remote_endpoint = boost::lexical_cast<std::string>(self->socket_.remote_endpoint());
+               boost::system::error_code re_ec;
+               auto re = self->socket_.remote_endpoint(re_ec);
```

Comment:
This implementation has been verified to fix the problem by the reporter. My concern with using the overloaded `async_accept` is that it would throw for `Transport endpoint is not connected` and it is unclear how to recover from that. My first go https://github.com/AntelopeIO/leap/pull/1175/commits/d3a00fcc4e1c7960aebeb81a144d820b46a00cc4 had that issue and caused nodeos to get stuck in an `accept: Already open` loop.

Reply:
sure, it's just clean up, you can decide. Or decide to do differently 4.0 vs main.

[The underlying syscall ](https://www.man7.org/linux/man-pages/man2/accept.2.html)returns the peer address atomically which I'm pretty sure asio uses for this use case of `async_accept()` (but I didn't confirm), making it impossible to fail when used in this manner. (asking for `remote_endpoint()` calls `getpeername`, which is where the race creeps in)

> My first go [d3a00fc](https://github.com/AntelopeIO/leap/commit/d3a00fcc4e1c7960aebeb81a144d820b46a00cc4) had that issue and caused nodeos to get stuck in an `accept: Already open` loop.

This makes some sense: you're calling `async_accept()` with.. a socket that's already open. Seems needed to `self->socket_.close()` in the error handler before reentering the accept loop. But I don't see this preconditioned documented anywhere.